homeassistant/components/madvr/coordinator.py

```diff
@@ -0,0 +1,56 @@
+"""Coordinator for handling data fetching and updates."""
+
+import logging
+
+from madvr.madvr import Madvr
+
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.update_coordinator import DataUpdateCoordinator
+
+_LOGGER = logging.getLogger(__name__)
+
+type MadVRConfigEntry = ConfigEntry[MadVRCoordinator]
+
+
+class MadVRCoordinator(DataUpdateCoordinator[dict]):
+    """My custom coordinator for push-based API."""
+
+    config_entry: MadVRConfigEntry
+
+    def __init__(
+        self,
+        hass: HomeAssistant,
+        config_entry: MadVRConfigEntry,
+        client: Madvr,
+        name: str,
+    ) -> None:
+        """Initialize my coordinator."""
+        super().__init__(
+            hass,
+            _LOGGER,
+            name="Madvr Coordinator",
+        )
+        self.entry_id = config_entry.entry_id
+        self.client = client
+        self.name = name
+        self.client.set_update_callback(self.handle_push_data)
+        _LOGGER.debug("MadVRCoordinator initialized")
+
+    async def _async_update_data(self):
+        """No-op method for initial setup."""
+        return
+
+    def handle_push_data(self, data: dict):
+        """Handle new data pushed from the API."""
+        _LOGGER.debug("Received push data: %s", data)
+        self.async_set_updated_data(data)
+
+    async def async_handle_unload(self):
```

Comment:
missing return type annotation

Reply:
I think you could even move this content to either `async_unload_entry` (because I think it makes sense since `__init__.py` creates the client, so it also has to take care of unloading the client). You could also make use of callbacks, an example:
https://github.com/home-assistant/core/blob/540da3cac6fef9d2fe691ead9718142129ed5832/homeassistant/components/nanoleaf/__init__.py#L80-L87